File: ChineseVoiceTranslator/ChineseVoiceTranslator/LanguageManager.swift
import Foundation
import Combine

enum AppLanguage: String, CaseIterable {
    case english = "en"
    case serbianCyrillic = "sr-Cyrl"
    case serbianLatin = "sr-Latn"

    var displayName: String {
        switch self {
        case .english: return "English"
        case .serbianCyrillic: return "Српски (ћирилица)"
        case .serbianLatin: return "Srpski (latinica)"
        }
    }

    var flag: String {
        switch self {
        case .english: return "🇬🇧"
        case .serbianCyrillic, .serbianLatin: return "🇷🇸"
        }
    }
}

class LanguageManager: ObservableObject {
    static let shared = LanguageManager()

    @Published var language: AppLanguage {
        didSet { UserDefaults.standard.set(language.rawValue, forKey: "appLanguage") }
    }

    init() {
        let saved = UserDefaults.standard.string(forKey: "appLanguage") ?? "en"
        language = AppLanguage(rawValue: saved) ?? .english
    }

    var s: Strings { Strings(language) }
}

File: flutter_app/lib/language_manager.dart
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'strings.dart';

enum AppLanguage { english, serbianCyrillic, serbianLatin }

class LanguageManager extends ChangeNotifier {
  static const _key = 'app_language';

  AppLanguage _language = AppLanguage.english;

  AppLanguage get language => _language;

  AppStrings get s => AppStrings(_language);

  // Returns the raw value sent to the backend for quiz evaluation language
  String get languageCode {
    switch (_language) {
      case AppLanguage.serbianCyrillic:
        return 'sr-Cyrl';
      case AppLanguage.serbianLatin:
        return 'sr-Latn';
      default:
        return 'en';
    }
  }

  Future<void> load() async {
    final prefs = await SharedPreferences.getInstance();
    final val = prefs.getString(_key);
    if (val != null) {
      _language = AppLanguage.values.firstWhere(
        (e) => e.name == val,
        orElse: () => AppLanguage.english,
      );
    }
    notifyListeners();
  }

  Future<void> setLanguage(AppLanguage lang) async {
    _language = lang;
    notifyListeners();
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_key, lang.name);
  }
}
